### PR TITLE
ONNX Loop operation support

### PR DESCRIPTION
**Pull request by lazarevevgeny**
Description: Added support for ONNX Loop operation in the MO. Added extractor for ONNX Min operation. Added support for ONNX ThresholdedRelu operation through decomposition. Fixed bug in the shape inference for ONNX Slice operation. Added ability to generate IR for TF models with "while" loops of some specific structure.

JIRA: 37683

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: partially covered in this PR + some transformations will be covered in the layer tests for ONNX Loop (41638)
* [x]  Framework operation tests: 41638
* [x]  Transformation tests
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: 38773
* [x]  Model Optimizer IR Reader check: IRs with Loop are not supported

Documentation:
* [x]  Supported frameworks operations list
* [x]  Supported **public** models list: N/A now new public model enabled
* [x]  New operations specification: already done
* [x]  Guide on how to convert the **public** model: N/A no new public model enabled
* [x]  User guide update: N/A no user visible change

### CODE REVIEW

**Review by achetver — APPROVED** [review]
Looks good to me.

**Review by mvafin — APPROVED** [review]

File: model-optimizer/extensions/ops/tensor_iterator.py [hunk 1]
```diff
@@ -364,6 +371,57 @@ def ti_type_infer(node):
...
+            assert ti_input_port.disconnected()
+            assert external_node_out_port.node.id not in ti_node.body
+            ti_input_port.connect(external_node_out_port)
+
+        ti_node.input_port_map.append({'axis': axis, 'stride': stride, 'part_size': part_size, 'start': start,
+                                       'end': end, 'external_port_id': ti_input_port.idx,
+                                       'internal_layer_id': internal_parameter['internal_layer_id']})
+
+    @staticmethod
+    def connect_body_output(ti_output_port: Port, internal_result: Node, external_node_input_ports: list = None,
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/tensor_iterator.py]: Moved to the Loop op.

File: model-optimizer/extensions/ops/tensor_iterator.py [hunk 2]
```diff
@@ -247,7 +247,8 @@ def substitute_ie_attrs(self, new_attrs: dict):
             'start',
             'stride',
             'end',
-            'part_size'
+            'part_size',
+            'purpose',
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/tensor_iterator.py]: Moved to a separate function

File: model-optimizer/extensions/ops/loop.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+from mo.front.common.partial_infer.utils import int64_array
+from mo.graph.graph import Node, Graph
+from mo.middle.passes.infer import partial_infer
+from mo.ops.const import Const
+
+
+class Loop(TensorIterator):
+    """
+    Loop layer that iterates over tensors and execute embedded sub-graph. The main difference from the TensorIterator is
+    that Loop operation performs implicit slicing of data using special input called "current_iteration". Also the Loop
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/ops/loop.py]: Fixed

File: model-optimizer/extensions/front/onnx/loop_ext.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        body_graph.graph.update(main_graph_attrs_copy)
+        loop_node['body'] = body_graph
+
+        # maps a tensor name to a node produced it and the node port: str -> (node_id, node_port)
+        data_nodes_map = {}
+        body_graph.graph['tensor_mapping'] = data_nodes_map  # save mapping for possible Loop inside the Loop
+
+        body_parameters = add_initializers_and_inputs_to_graph(body_graph, body_graph_proto, data_nodes_map)
+
+        external_edges = []  # (src_node, src_out_port), dest_body_parameter_node
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/onnx/loop_ext.py:79]: Yes, there are some. I plan to unify the reader when work on Loop inside Loop support. For now, let's keep this partial copy-paste.

File: model-optimizer/extensions/front/onnx/LoopNormalize.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes(op='Loop'):
+            self.normalize_body_graph(node)
+
+    @staticmethod
+    def normalize_body_graph(loop_node: Node):
+        loop_name = loop_node.soft_get('name', loop_node.id)
+        # connect "trip count" input if it is not connected with default value "Infinity" (-1)
+        if 0 not in loop_node.in_ports() or loop_node.in_port(0).disconnected():
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/onnx/LoopNormalize.py]: Thanks. Fixed.

File: model-optimizer/extensions/front/ThresholdedReluDecomposition.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    ThresholdedRelu(x, alpha) = Mul(x, Cast(Greater(x, alpha), type=float))
+    """
+    enabled = True
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes(op='ThresholdedRelu'):
+            name = node.soft_get('name', node.id)
+
+            greater = create_op_with_const_inputs(graph, Greater, {1: node.alpha})
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/front/ThresholdedReluDecomposition.py]: We rely on a transformation of the Const type in the back phase.
Anyway I changed it to Const with shape [1] to be more plugins friendly ([1] is broadcastable to any shape anyway so this should be a safe transformation).

File: model-optimizer/extensions/middle/TensorIteratorMerge.py [hunk 7]
```diff
@@ -120,21 +120,22 @@ def pattern():
     def replace_pattern(graph, match: dict):
         # Here we will found all parts of TI: condition, inputs/outputs, back edges, body and create TensorIterator Op
         # and make all checks needed for TensorIteator work
-        cond_data = match['condition'].out_node(0)
-        time_data = match['condition'].out_node(1) if len(match['condition'].out_nodes()) > 1 else None
+        cond_data = match['condition'].out_node(0) if not match['condition'].out_port(0).disconnected() else None
```
[the remark this answers is not in the document]
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/TensorIteratorMerge.py:123]: This change allow to enable some TF "while" loop cases. These changes are just relaxing of some conditions, so they should not break anything that worked before.

**Review by jane-intel — APPROVED** [review]

File: model-optimizer/extensions/front/onnx/loop_ext.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+            # create an NX node
+            id = body_graph.unique_id(node_id(pb_node))
+            body_graph.add_node(id, pb=pb_node, kind='op')
+
+            # add incoming edges based on data_nodes_map
+            for dst_port, inp in enumerate(pb_node.input):
+                # should add edge inp --> id
+                if inp not in data_nodes_map:
+                    if inp == '':
+                        # input is omitted; most likely it corresponds to an optional input for an operator
```
Comment (itikhono) [inline model-optimizer/extensions/front/onnx/loop_ext.py:92]: I guess we need to add some logic here to make Const(true) -> conditional output subgraph in the body

Comment (lazarevevgeny) [reply model-optimizer/extensions/front/onnx/loop_ext.py:92]: Will be a part of another PR.

**Review by itikhono — APPROVED** [review]

**Pull request merged**